bpf-notifier/notifier.c

```diff
@@ -0,0 +1,139 @@
+// SPDX-License-Identifier: (LGPL-2.1 OR BSD-2-Clause)
+/* Copyright (c) 2023 Red Hat */
+
+#include <argp.h>
+#include <signal.h>
+#include <stdio.h>
+#include <time.h>
+#include <unistd.h>
+#include <sys/resource.h>
+#include <bpf/libbpf.h>
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <net/if.h>
+#include <linux/if_link.h>
+#include <linux/if_ether.h>
+#include <bpf/bpf.h>
+#include <netinet/in.h>
+#include <netinet/ether.h>
+#include <arpa/inet.h>
+#include <string.h>
+
+#include "notifier.skel.h"
+#include "notifier.h"
+
+static struct env {
+	int interval;
+	bool verbose;
+} env = { 1, 0 };
+
+const char *argp_program_version = "notifier 0.0";
+const char *argp_program_bug_address = "<donald.hunter@redhat.com>";
+const char argp_program_doc[] =
+"BPF notifier prototype.\n"
+"\n"
+"USAGE: ./notifier [-v]\n";
+
+static const struct argp_option opts[] = {
+	{ "interval", 'i', "seconds", 0, "Interval between reports" },
+	{ "verbose", 'v', NULL, 0, "Verbose debug output" },
+	{},
+};
+
+static error_t parse_arg(int key, char *arg, struct argp_state *state)
+{
+	switch (key) {
+	case 'i':
+		env.interval = atoi(arg);
+		break;
+	case 'v':
+		env.verbose = true;
+		break;
+	case ARGP_KEY_ARG:
+		argp_usage(state);
+		break;
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return 0;
+}
+
+static const struct argp argp = {
+	.options = opts,
+	.parser = parse_arg,
+	.doc = argp_program_doc,
+};
+
+static int libbpf_print_fn(enum libbpf_print_level level, const char *format, va_list args)
+{
+	if (level == LIBBPF_DEBUG && !env.verbose)
+		return 0;
+	return vfprintf(stderr, format, args);
+}
+
+static volatile bool exiting = false;
+
+static void sig_handler(int sig)
+{
+	exiting = true;
+}
+
+static int process_event(void *ctx, void *data, size_t len) {
+	struct notifier_event *event = data;
+
+        printf("notifier event id=%lu, mac=%s, name=%s\n", event->val,
+               ether_ntoa((struct ether_addr*)event->mac), event->name);
+        return 0;
+}
+
+int main(int argc, char **argv)
+{
+	struct notifier_bpf *skel;
+	int err;
+
+	/* Parse command line arguments */
+	err = argp_parse(&argp, argc, argv, 0, NULL, NULL);
+	if (err)
+		return err;
+
+	libbpf_set_strict_mode(LIBBPF_STRICT_ALL);
+	/* Set up libbpf errors and debug info callback */
+	libbpf_set_print(libbpf_print_fn);
+
+	/* Cleaner handling of Ctrl-C */
+	signal(SIGINT, sig_handler);
+	signal(SIGTERM, sig_handler);
+
+	/* Load and verify BPF application */
+	skel = notifier_bpf__open();
```

Comment:
this is probably a stupid question :( where are the notifier_bpf__open/load/attach/destroy (functions) defined?

Reply:
or probably something I'm not understanding with the fentry progs :(

Reply:
Good question. The `notifier_bpf__*` functions get generated by `bpftool gen skeleton` from `notifier.bpf.o` during the build. The generated file is `.output/notifier.skel.h`. The generated skeleton provides all the convenience wrappers around libbpf usage.